ipa-core/src/test_fixture/hybrid_event_gen.rs

```diff
@@ -0,0 +1,241 @@
+use std::num::{NonZeroU32, NonZeroU64};
+
+use rand::Rng;
+
+use super::hybrid::TestHybridRecord;
+
+#[derive(Debug, Copy, Clone)]
+#[cfg_attr(feature = "clap", derive(clap::ValueEnum))]
+pub enum ConversionDistribution {
+    Default,
+    LotsOfConversionsPerImpression,
+    OnlyImpressions,
+    OnlyConversions,
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "clap", derive(clap::Args))]
+pub struct Config {
+    #[cfg_attr(feature = "clap", arg(long, default_value = "1000000000000"))]
+    pub num_events: NonZeroU64,
+    #[cfg_attr(feature = "clap", arg(long, default_value = "5"))]
+    pub max_conversion_value: NonZeroU32,
+    #[cfg_attr(feature = "clap", arg(long, default_value = "20"))]
+    pub max_breakdown_key: NonZeroU32,
+    #[cfg_attr(feature = "clap", arg(long, default_value = "10"))]
+    pub max_convs_per_imp: NonZeroU32,
+    /// Indicates the distribution of impression to conversion reports.
+    #[cfg_attr(feature = "clap", arg(value_enum, long, default_value_t = ConversionDistribution::Default))]
+    pub conversion_distribution: ConversionDistribution,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self::new(1_000, 5, 20, 10)
+    }
+}
+
+impl Config {
+    /// Creates a new instance of [`Self`]
+    ///
+    /// ## Panics
+    /// If any argument is 0.
+    #[must_use]
+    pub fn new(
+        num_events: u64,
+        max_conversion_value: u32,
+        max_breakdown_key: u32,
+        max_convs_per_imp: u32,
+    ) -> Self {
+        Self {
+            num_events: NonZeroU64::try_from(num_events).unwrap(),
+            max_conversion_value: NonZeroU32::try_from(max_conversion_value).unwrap(),
+            max_breakdown_key: NonZeroU32::try_from(max_breakdown_key).unwrap(),
+            max_convs_per_imp: NonZeroU32::try_from(max_convs_per_imp).unwrap(),
+            conversion_distribution: ConversionDistribution::Default,
+        }
+    }
+}
+
+pub struct EventGenerator<R: Rng> {
+    config: Config,
+    rng: R,
+    in_flight: Vec<TestHybridRecord>,
+}
+
+impl<R: Rng> EventGenerator<R> {
+    #[allow(dead_code)]
+    pub fn with_default_config(rng: R) -> Self {
+        Self::with_config(rng, Config::default())
+    }
+
+    /// # Panics
+    /// If the configuration is not valid.
+    #[allow(dead_code)]
+    pub fn with_config(rng: R, config: Config) -> Self {
+        let max_capacity = usize::try_from(config.max_convs_per_imp.get() + 1).unwrap();
+        Self {
+            config,
+            rng,
+            in_flight: Vec::with_capacity(max_capacity),
+        }
+    }
+
+    fn gen_batch(&mut self) {
+        match self.config.conversion_distribution {
+            ConversionDistribution::OnlyImpressions => {
```

Comment:
I think we should test these branches too if we plan to offer them. Alternatively we could remove it and add it with tests when we need them